test/functional/interface_ipc.py

```diff
@@ -204,13 +208,34 @@ async def async_routine():
             res = await template.result.submitSolution(ctx, block.nVersion, block.nTime, block.nNonce, coinbase.serialize())
             assert_equal(res.result, False)
             self.log.debug("Submit a valid block")
+            # The remote template block will be mutated, capture the original:
+            remote_block_before = await self.parse_and_deserialize_block(template, ctx)
             block.nVersion = original_version
             block.solve()
+
+            self.log.debug("First call checkBlock()")
             res = await mining.result.checkBlock(block.serialize(), check_opts)
             assert_equal(res.result, True)
+
+            self.log.debug("Submitted coinbase must include witness")
+            assert_not_equal(coinbase.serialize_without_witness().hex(), coinbase.serialize().hex())
+            res = await template.result.submitSolution(ctx, block.nVersion, block.nTime, block.nNonce, coinbase.serialize_without_witness())
+            assert_equal(res.result, False)
+
+            self.log.debug("Submit again, with the witness")
             res = await template.result.submitSolution(ctx, block.nVersion, block.nTime, block.nNonce, coinbase.serialize())
             assert_equal(res.result, True)
-            assert_equal(self.nodes[0].getchaintips()[0]["height"], current_block_height + 1)
+
+            self.log.debug("The above submitBlock() mutated the template's block")
+            remote_block_after = await self.parse_and_deserialize_block(template, ctx)
+            assert_not_equal(remote_block_before.serialize().hex(), remote_block_after.serialize().hex())
+
+            self.log.debug("Block should propagate")
+            assert_equal(self.nodes[1].getchaintips()[0]["height"], current_block_height + 1)
+            # Stalls if a regression causes submitBlock() to accept an invalid block:
+            self.sync_all()
+            assert_equal(self.nodes[0].getchaintips()[0], self.nodes[1].getchaintips()[0])
```

Comment:
If you apply the patch in the description and also drop the `sync_all` call, you'll see that the second node sees the block as `headers-only`.